time estimates for plans

awalter-bnl opened the pull request:
This is part of a system that allows for time estimation for plans. It implements a telemetry and time estimation process at the ophyd object level ([NSLS-II/Ophyd PR # 567](https://github.com/NSLS-II/ophyd/pull/567)) to estimate the time to make a 'cmd' on an object. 'cmd' is the commands past in the plan messages.

An est_time('plan') function, similar to summarise_plans('plan'), then steps through each message in order to estimate the time to perform each step and hence the entire plan. 

This fixes issue #713 when combined with [NSLS-II / Ophyd PR # 567](https://github.com/NSLS-II/ophyd/pull/567)

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +114,276 @@ def check_limits(plan):
+        out_est_time = [0,0]
+        if method == 'sum':
+            out_est_time[0] = est_time_1[0] + est_time_2[0]
+            out_est_time[1] = est_time_1[1] + est_time_2[1]
```
Shouldn't the uncertainties be combined in a way that accounts for [propagation of errors](https://en.wikipedia.org/wiki/Propagation_of_uncertainty#Example_formulas)?

awalter-bnl replied:
These are not 'errors' but standard deviations. There are many ways to combine errors, this is just the simplest approach which works on a very general level. I am happy to discuss other approaches.

danielballan replied:
Yes, I mean “error” as in “standard error” — standard deviation. Possible clash of continental terms?

I think that adding them in quadrature is understood to be the best general method when the sources of deviation are many and independently distributed (central limit theorem).

awalter-bnl replied:
OK will look into this.

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +114,276 @@ def check_limits(plan):
+        out_est_time : list.
+            The combined est_time/std_dev tuple.
+        """
+        out_est_time = [0,0]
```
This might be a good place to use a `namedtuple`, which would support both the normal tuple usage (item access by position) _and_ access by name, such as `out_est_time.value` and `out_est_time.uncertainty`.

awalter-bnl replied:
I will look into this, I wasn't aware of 'namedtuples'.

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +114,276 @@ def check_limits(plan):
+
+
+    def obj_est_time(msg, val_dict):
+        """
```
Functions that mutate their inputs should document prominent that they do so. It can be useful, and seems useful in the case, but it can also be _very surprising_ so it needs to be clearly flagged.

awalter-bnl replied:
I thought I had clearly flagged this, I will make it clearer in the doc string.

awalter-bnl replied:
Not sure exactly how to do this as a class, but I will think about it. Is the idea that an instance of the 'class' will have the dictionary as an attribute?

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +114,276 @@ def check_limits(plan):
+                if msg.command == 'unstage':
+                    val_dict['trigger'][msg.obj.name] = 0
+                elif msg.command == 'trigger':
+                    if msg.obj.name in list(val_dict['trigger'].keys()):
```
Removing `list` here would save from making a needless copy.

prjemian replied:
Isn't this identical:

```
if msg.obj.name in val_dict['trigger']:```

awalter-bnl replied:
OK will do, for some reason I thought that this wouldn't work.

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +114,276 @@ def check_limits(plan):
+
+        elif method == 'max':
+            if est_time_1[0] < est_time_2[0]:
+                out_est_time = est_time_2 
```
Combining uncertainties here is tricky. Maybe a soft maximum function is the right approximation?

awalter-bnl replied:
Here it is not so tricky, because we are assuming that the time taken is that of the larger of the two. Therefore the error should also be the error associated with the larger of the two.

awalter-bnl replied:
OK so I hadn't considered that in some cases the standard deviation could increase the time of one above the maximum time. This seems easy enough to work around, I will make this change too.

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +114,276 @@ def check_limits(plan):
+
+
+    #Define some variables used in the following.
+    val_dict = {'set':{}, 'trigger':{} } #this holds information on the updated values.
```
Now for a signature naming-related nitpick! :-D

I don't find `val_dict` particularly descriptive. I usually see variables with the kind role with `state` in the name, which also hints that this is is something we are going to mutating a lot. Maybe just name it `state`?

prjemian replied:
`state` means *so* many different things and is likely to already have implementation in cases where time estimates would be a valuable add-on.  At a loss how to make this name more specific (and less likely to conflict)-- ideas?

awalter-bnl replied:
I too am not so happy with val_dict, I will think about what else it can be but don't like state for the same reasons as @prjemian

prjemian replied:
`msg_info` ?  Seems concise, descriptive for its context, and not likely in use

tacaswell replied:
This is pretty well scoped (it is not actually publicly exposed at all right now!) so I am not worried about collisions with other uses of `state`.  I like `state` quite a bit here, but I think that this is a candidate for a class (like the RE) to handle this via the visitor pattern (ex https://en.wikipedia.org/wiki/Visitor_pattern ).

prjemian replied:
ok with naming.  Agree there is more complexity here; a class is warranted.

awalter-bnl replied:
I quite like msg_info, but maybe plan_info is better as it is keeping track of what items have been 'modified' in the plan up to this point. As I mentioned above I am not exactly sure how this would work as a class, but will think about it.

awalter-bnl replied:
To try to clarify the use of val_dict, essentially to know how long it takes to 'move' a device it is necessary to know the starting point. As this can be modified by an earlier message in the plan when stepping through messages without performing the action I need to keep track of any previous changes to this that would have been caused by the plan itself. There are also other changes (settle_time etc) that may be 'changed' during a plan, and can change the time estimate, that need to be tracked.

prjemian replied:
Complex numbers are probably the best example I can think of.  Special math operations for them.  This `msg_info` (or is it now `plan_history`) class could combine the special operations with the historical info such as a database.  Not sure I well understand what is required.  Assume the doc strings will describe in some detail.

awalter-bnl replied:
I thin the docstrings already describe pretty well, if not let me know and I will update them. Essentially, as an example, think of an 'area detector' detector. To a first approximation the est_time for a triggering can be written as est_time = settle_time + num_images * acquire_time.

In a plan we could easily change the acquire_time part of the way through the plan. If this happens then we would need this 'updated' value in order to estimate the time for any following triggers of this detector in the plan. As the generator function does not actually perform this change, it needs to be 'tracked' which is what I use val_dict(name change pending) for.

Similar things occur for other actions (set for example by definition changes values). Hopefully this is a little clearer. I think I have a better way to do this using classes, I will write it up and see how it goes.

awalter-bnl commented:
@danielballan, relating to your big picture thoughts. 

1. I am happy to move the smaller function from the larger one, however I wanted to avoid polluting the namespace with a bunch of smaller functions. What do you think of _function to prevent this instead?

2. I agree we should look into this, but leave it for a future PR.

3. Each ophyd obj now already has an attribute function that provides this tuple for a given command, so I don't see the benefit of this approach. The 'granularity' you are requesting will then be provided by all the sub-functions that I will pull out.

awalter-bnl commented:
No Problem, on any of these points. I will make the changes. For number 3 I will think about the best way to approach this. Due to the complications of 'groups' which are performed in parallel I am not sure it is a simple as you predict. If you send in a series of msgs with a 'group' in the middle do you want it to return the time for each msg or 1 for the group and 1 for each msg inside the group? I am guessing the later but am not totally sure how you plan to arrange the progress bar for groups.

awalter-bnl commented:
I have pushed a number of commits related to the issues described above. I think I have got all of the requests, with the exception of the move to put the telemetry recording on status objects. This is because I am having some issues understanding how status objects work from the code, I do intend to make this change once that is clearer to me however. Notes on the changes are below, each numbered item should have a corresponding commit for ease of following.

1. I turned the simulator est_time into a class 'EstTimeSimulator' and an instantiation 'est_time' because:
     - This allows easy access to the sub functions as requested by @danielballan 
     - This allows for a class definition of val_dict, which no longer has to be passed in and out of sub  
     classes, as requested by @danielballan  and @tacaswell  
2. Modifed EstTimeSimulator (bluesky), sim.py (ophyd), device.py (ophyd) and  EstTime (ophyd) to change how val_dict is used and its name.
    - in EstTimeSimulator (bluesky) it is a class attribtue self._plan_history.
    -  This means it is not longer passed into and returned from functions as requested by  
    @danielballan  and @tacaswell  and has a new name as requested by @danielballan and @prjemian.
    - in EstTime (ophyd), device.py (ophyd), sim.py (ophyd) it is an input only variable called 
    plan_history.
3. Changed the stats and est_time to namedtuples
    - Also updated some of the names for these to make it clearer.
4. remove telemetry and est_time on read.
    - At the request of @danielballan   
5. changed the method for combining std_dev's.
    - At  the request of @danielballan   
6. Added a new function obj.est_time.msg_est(plan)
    - This was done at the request of @danielballan  to help with a plan progress bar.
    - Contrary to @danielballan's request I have not made this a 'base' function of the main 
    obj.est_time(plan) simulator function as I think this would add a second iteration through 
    the list of messages which will add an undue processing time burden on obj.est_time(plan).

danielballan commented:
Description looks good. Will dig into code as soon as I can. Happy to help walk through the Status code (and document it better in the process!)

awalter-bnl commented:
Thanks @Daniel Allan , I will take you up on the offer of a discussion of status objects. Perhaps next week when you aren’t in Chicago.

danielballan commented:
> For number 3 I will think about the best way to approach this. Due to the complications of 'groups' which are performed in parallel I am not sure it is a simple as you predict. If you send in a series of msgs with a 'group' in the middle do you want it to return the time for each msg or 1 for the group and 1 for each msg inside the group? I am guessing the later but am not totally sure how you plan to arrange the progress bar for groups.

I think that it should work like this:

```
# msg ->  est_time
Msg('set', motor1, ..., group='A')  -> 0
Msg('set', motor2, ..., group='A')  -> 0
Msg('set', motor3, ..., group='B')  -> 0
Msg('wait', 'A')  -> max(motor1.est_time.set(...), motor2.est_time.set(...))
Msg('wait', 'B')  -> motor3.est_time.set(...)
```

This is an accurate representation of how long each message will take to process: 'set' messages are instantaneous for RE; only the 'wait' messages take time.

awalter-bnl commented:
Thanks @d
at present I do the former not the later, very similar to what you suggest, without separating by 'groups'. I will work on changing this.

awalter-bnl commented:
I have updated this so that the est_time.group_eta function can handle multiple parallel 'groups'.

danielballan commented:
This class has ["two methods one of which is `__init__`"](https://www.youtube.com/watch?v=o9pEzgHorH0) -- plus some utility methods that are tucked into the class for scoping / namespace-decluttering reasons. Let's refactor it as a _function_ plus utility functions. The scoping / namespace-decluttering is already provided by the module itself:

```py
from bluesky import est_time
est_time.some_function(...)
```

awalter-bnl commented:
Have made the change requested by @danielballan (removing EstTimeSimulator class) and tested.

Sorry for forgetting we had discussed this earlier, for posterity I am still not sure why this is better than the class structure. However as it doesn't prevent any of the aims of this PR being realized I ahve made the change.

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +121,401 @@ def check_limits(plan):
+    beginning of the scan or since an 'unstage' event that the device has been triggered.
+
+"""
+#The commands that start/end a 'run'.
```
Most -- maybe all -- of these lists strike me as unneeded obfuscation. There is only one command that starts are run, `'open_run'`, and it is not likely to change in the future. Why not just type `'open_run'` directly where it's needed?

It occurs to me you might intend this as path to registering custom user commands (message types) with this system. User commands are somewhat limited: a user can't add a comment that interacts with private RunEngine state, including opening/closing runs or doing anything with groups. It is safe to hard-code these command names, for the foreseeable future.

awalter-bnl replied:
I am not convinced that hard-coding these is a good idea. My strong preference is to leave it as is in case things change in the future to make it easier to change.

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +121,401 @@ def check_limits(plan):
+_flyer_start_cmds = ['kickoff']
+#create a reference to a dictionary called plan_history to capture set changes during the 
+#plan for later modification.
+_plan_history = {'set':{}, 'trigger':{}}
```
This shouldn't be a module global; it should be defined internally to `est_time` and passed explicitly into `obj_est`, `group_est`, etc. A mutable global creates the possibility for unintended interactions. Can be useful, but no reason to open Pandora's box in this case.

awalter-bnl replied:
This was done explicitly because @tacaswell was very against me passing a variable in to a function that was modified in the function and the returned. So a decision needs to be made between @tacaswell and @danielballan about which approach to use as the requests are mutually exclusive.

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +121,401 @@ def check_limits(plan):
+        A list containing a msg/est_time/std_dev named tuple for each message in the plan.
+    '''
+    
+    out_time = []
```
Rather than accumulating the messages in a list and then returning the list, this can be implemented as generator, yielding out one estimate at a time. This makes it more flexible. If the caller _wants_ a list, then can of course call `list(msg_est(plan))`.

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +121,401 @@ def check_limits(plan):
+
+    Updated Parameters
+    ------------------
+    _plan_history : dict.
```
This should be promoted to a parameter and passed in by all callers.

awalter-bnl replied:
As mentioned above, this was done to avoid having the functions take this in as an argument and return a possibly modified version.

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +121,401 @@ def check_limits(plan):
+_plan_history = {'set':{}, 'trigger':{}}
+
+def est_time(plan, print_output = True):
+    '''The call method.
```
Remove 'the call method' now that this is a function.

awalter-bnl replied:
fixed

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +121,401 @@ def check_limits(plan):
+    This method takes in a plan, and an optional print_output kwarg, and returns an estimate for 
+    the time to complete for the plan, in addition it returns an time estimate for any subplans. 
+
+    PARAMETERS
```
Sphinx expects the capitalization `Parameters`.

awalter-bnl replied:
fixed

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +121,401 @@ def check_limits(plan):
+        Indicates if the return values should also be printed to the command line in a human 
+        readable way, default value is 'True'.
+
+    RETURN PARAMETERS
```
Sphinx expects this to be spelled `Returns`

awalter-bnl replied:
fixed

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +121,401 @@ def check_limits(plan):
+            #The if-elif section below is used to track how many 'triggers'/'sets' have 
+            #occured since the last 'unstage'/start of plan.
+            if msg.command == 'unstage':
+                _plan_history['trigger'][msg.obj.name] = 0
```
`del _plan_history['trigger'][msg.obj.name]` might be a clearer way of indicating "not applicable" than `0`.

awalter-bnl replied:
This is set as zero as the 'value' is the number of triggers since the last 'unstage' not just an indication if it is staged. I did it this way in preparation for the custom devices like at CSX where some staging is done in the first trigger.

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +121,401 @@ def check_limits(plan):
+            object_est = obj_est(msg)
+            out_time = combine_est(out_time, object_est)
+
+        if print_output == True:
```
`if print_output` is sufficient.

awalter-bnl replied:
fixed

awalter-bnl commented:
I disagree @danielballan , they way it is written it absolutely takes into account sets/groups that occur outside of runs. I have chosen to also provide the time for the 'runs', as a separate list to the total time, to give the users a better understanding of which part of the plan might be taking to long and to help them organise the plans easier as the 'macro' type plans that most users will write will contain several 'runs' as their basic building block. I have treated 'runs' in the same way as 'groups', in that the times are separated out and added to the total time for the plan once the end of the 'run'/'group' has been found.

To clarify the total 'plan' time is not just the sum of the times for each 'run' but the sum of everything that occurs inside 'runs' AND everything that occurs between 'runs' and so can be longer than the sum of all 'runs'. I have tested the code to make sure this works by adding 'yield from mv....' and 'yield from wait..' lines between 'yield from scan......' lines to a test plan I was using. 

I have specifically NOT written the code so that it first runs 'msg_est' and then go back through this and sort into 'runs', 'groups' etc. because this means counting through the messages more than once, and hence doubles the time overhead of calling est_time.

danielballan commented:
> I have specifically NOT written the code so that it first runs 'msg_est' and then go back through this and sort into 'runs', 'groups' etc. because this means counting through the messages more than once, and hence doubles the time overhead of calling est_time.

Agreed that needing to iterate through the plan twice would be a bad design. I think this can be done in a single pass. Let me sketch out a little more specifically the design I had in mind with [this comment](https://github.com/NSLS-II/bluesky/pull/1048#issuecomment-396098919).

```py
# These constants should be revised using the test stand.
COMMUNICATION_TIME = (0.1, 0.1)  # It takes 0.1 +/- 0.1 to send a command to the IOC.
INTERNAL_STATE_TIME = (3e-5, 3e-5)  # It takes ~3e-5 seconds to process msgs like 'checkpoint'.

def est_time(plan):
    "Takes in a plan and yields out (time, std_err) per message."
    history = {}
    groups = defaultdict(list)
    for msg in plan:
        command = msg.command
        # Case 1: commands that generate a status object
        if command in ('set', 'trigger', 'kickoff', 'complete'):
            # TODO something involving history here...
            t = getattr(msg.obj.est_time, command)(...)
            groups[msg.kwargs.get('group')].append(t)
            yield COMMUNICATION TIME
        # Case 2: waiting on a status object cached by Case 1
        elif msg.command == 'wait':
            times = groups.pop(msg.kwargs.get('group'))
            t, std_err = combine(times, 'max')
            # TODO Update other groups (subtract t).
            yield t, std_err
        # Case 3: commands that communicate with the IOC but are quick
        elif msg.command in ('read', 'describe'):
            yield COMMUNICATION_TIME
        # Case 4: TODO 'collect' -- should scale with number of data points
        # Case 5: temporary special-case: these _should_ return a status object but don't yet
        elif command in ('stage', 'unstage'):
            yield (0, 0)  # temporary, unrealistic
        # Case 6: all other commands that do not involve network communication, are fast
        else:
            yield INTERNAL_STATE_TIME

def est_time_per_run(plan):
    "Takes a plan and yields out total (time, std_err) per run."
    times = []  # Accumulate times in current run.
    in_run = False  # Tracks whether we are currently inside a run.
    a, b = itertools.tee(plan)
    for msg, est in zip(a, est_time(b)):
        if msg == 'open_run':
            in_run = True
        if msg == 'close_run':
            in_run = False
            yield combine(times, 'max')
        if in_run:
            times.append(est)
```

Things I like about this general design:

* We obviate the discussion cited above about a global dict passing around mutable state. The state lives as a local var inside `est_time`. It is not shared between functions.
* We can imagine several reductions/aggregations following the pattern of `est_time_per_run`. Someone could write `est_time_total`, `est_time_per_sample` or `est_time_per_point` using the same pattern, and all the "tricky" logic is safely contained in `est_time`.

danielballan commented:
One clarification that occurred to me I should state: Note that because `est_time` is a _generator_, we aren't looping through the messages twice. We're capturing output _as it comes out_ and tallying. It's layered, but it's not doubling the loop.

awalter-bnl commented:
OK so I am slowly becoming convinced, I see two issues to be resolved first however:

ISSUE 1.
It would need another function on top of this to return the times in a useful way for users (e.g. a 'print_summary' style wrapper for printing or some other method for exporting the data to a GUI). 
- This may not be a major issue but want to get feedback on what people think regarding this as the API. 
- What do people think of a single API with boolean keywords used to determine what is printed?
    - eg. est_time_plan(plan, total = true, runs = true, .......)
    **NOTE: I am using the 'summarize_plan' naming convention here, what do people think about this?

ISSUE 2
I think that 95% of the time the users will want both the total estimated time AND the time per run. In your setup this would still require stepping through the message list twice, once when running est_time_per_run(plan) and once when running est_time_total(plan). 
- I see two options to get around this:
    - both would require the 'yields' to give 3 values 'name', est_time and std_dev with 'name' being
    something like 'msg x', 'group y' or 'run z' which can be used in the 'print' wrapper described above.

    1. Make est_time_per_run also yield the time estimate for msg's between the runs, including the 
    'name' parameter described above. Then we can simply pass est_time_per_run(est_time(plan)) as 
    the argument of est_time_total to output both. This relies on the fact that both est_time and 
    est_time_per_run are generators with the same structure for the output and so either can be the
    input of est_time_total. The 'name' can then be used to distinguish what type of 'thing' the time 
    estimate relates too. 

     2. we could make est_time_per_run(plan) also return the total time for the plan as the last 'yield'. If 
    we gave it the name 'total' we could distinguish it from those with the name 'run'. 

FURTHER THOUGHTS
In principal I like number 2 as it is explicit, but not strongly enough to insist. Number 1 has the benefit that it is easy to nest any future reductions in which could be useful.

I think that adding the 'name' as part of each yield would make a generic 'print' wrapper easy (the first issue I mentioned). It would just need to print the following on each yield:
`print (f.'{name}: est time = {est_time}, std_dev = {std_dev}')`

danielballan commented:
> What do people think of a single API with boolean keywords used to determine what is printed?
eg. `est_time_plan(plan, total = true, runs = true, .......)`
**NOTE: I am using the 'summarize_plan' naming convention here, what do people think about this?

The job of `summarize_plan` is to remove information. That is, when`list(plan)` is too much information, `summarize_plan` is there to construct an incomplete but often sufficient and nicer-to-read representation. It does not add any new information, so there is not much need to access its output programmatically -- might as well just dump to the screen.

The job of `est_time`, in contrast, is to provide additional information. That information should be programmatically accessible so that it can be reused by other components (GUIs, user-defined functions, etc).

This usage seems user-friendly to me, and unlike prints it is also programmatically useful:

```py
>>> est_time_per_run(plan)
[EstTime(time=10, std_err=1), EstTime(time=5, std_err=0.3), EstTime(time=10, std_err=1)]
```

If I want the total I can do this:

```py
>>> est_time_total(plan)
EstTime(time=20, std_err=2.3)
```

or this:

```py
>>> sum([item.item for item in est_time_per_run(plan)])
20
```

or this:

```py
>>> combine(est_time_per_run(plan))
EstTime(time=20, std_err=2.3)
```

As an additional convenience, I would not object to some high-level function that prints something like:

```
Total Estimated time: 20 +/ 2.3
Per run breakdown:
(etc.)
```

> I think that 95% of the time the users will want both the total estimated time AND the time per run. In your setup this would still require stepping through the message list twice, once when running est_time_per_run(plan) and once when running est_time_total(plan).

I think you are right that that is the common case and that we need a robust solution for it that ideally does not require iterating through the messages twice.

It just occurred to me that, when we extend the RunEngine to make it possible to have multiple runs open concurrently as we plan to do someday, the relationship between total time and time during and between runs will not be straightforward. Runs can overlap. We should try to avoid making the 'one run at a time' assumption in here to avoid future pain. Nonetheless the time 'per run' will remain a useful concept for the common case of non-overlapping runs.

Here's a proposal that solves your problem ("Users commonly want time per run and total time") and the problem I foresee ("Time per run is ill-defined.") I think `est_time_per_run` -- or whatever we call it -- should report the _cumulative_ `EstTime` each time a run is closed. This is well-defined even if runs overlap, it incorporates any time spent between runs, and tells the user what they probably want to know: "How long until I can see all of dataset X?" If we want to include the time spent after the last run closes, we can issue one last message at the end.

awalter-bnl commented:
So I want to push back a little here, and it is something that I was also considering for the group_est time as well as the potentially overlapping 'runs'.

I think the idea of a 'cumulative est_time' is a really good idea, but I think it would be beneficial if we extended the output tuple to include it AND the time for the run/group defined by 'name'. This would result in a named tuple with the values ('name', 'cumulative estimate', 'section?? estimate'). In this case I imagine that 'cumulative estimate' and 'section?? estimate' are themselves named tuples with the values ('est_time', 'std_dev'). (I am not convinced of the name 'section' to relate to the 'group', 'run' or 'non-groupable action' that is referenced by name but don't have a better option). 

The reason I like this is, as you say, cumulative est_time outputs when a particular part of the scan is done and by default provides the total est_time at the end. However there is still some information that is lost. By also including the total time for a 'group'/'run' or 'non-groupable action' then we are not losing any time information.

As a use case I could imagine the extra info of the total 'plan'/ 'group' or 'non-groupable action' times would be useful for a user trying to determine which part of a 'macro-like plan' is taking the longest and where the best point to change things to reduce the time is. I could also imagine using this extra info to make a 'visual representation', similar to a schedule bar-graph, showing when different 'sections' start and stop throughout the plan which would be a neat and useful GUI addition that can help with scheduling 'macro-like plans' with many runs inside.

What are everyones thoughts?

danielballan commented:
> I think the idea of a 'cumulative est_time' is a really good idea, but I think it would be beneficial if we extended the output tuple to include it AND the time for the run/group defined by 'name'. 

Can you give an example of what this would look like? I'm not clear on where the 'name' come from.

awalter-bnl commented:
The 'name' is really just to distinguish between like actions/groups/runs. For example a run might be
for a run:
`('run_1', (cumulative_est_time, cumulative_std_dev), (section_est_time, section_std_dev))`
for a group:
`('group_'+msg.group, (cumulative_est_time, cumulative_std_dev), (section_est_time, section_std_dev))`
for a non-groupable action:
`(msg.cmd+'_1', (cumulative_est_time, cumulative_std_dev), (section_est_time, section_std_dev))`

The aim of this is to produce an output something like:
`$ list[est_time_per_run(plan)]
[('group_A', ........), ('run_1', .......), ('sleep_1', ........), ('group_B', .........), ('run_2',........), .................`

Does this make it a little clearer?

danielballan commented:
I think we agree on:

* We need the low-level generator (along the lines of my `est_time` sketch) to provide maximally-granular information -- time per `Msg` -- such as would be suitable to support a progress bar.
* We need a high-level generator (cumulative, perhaps) to provide a reduced/simplified view for human consumption.

I think you are proposal an intermediate variant, which includes a level of detail somewhere between these two. I think at this point we should consider moving to a visual representation rather than a text-based one -- as you proposed above. A intermediate text-based representation looks like it will be quite verbose and hard to parse for realistic plans. But a visualization, whether a simple static plot or an interactive GUI, can be derived from the low-level generator, and this lets us avoid trying to mind-read which things the user doesn't want to see ("details") and which things are important for them to see that day. It also avoids the need to try to infer "sections" from what is technically a flat list.

danielballan commented:
I do think it's useful to consider other reduced representations. I mentioned `est_time_per_point` above -- having in mind the time per Event.

danielballan commented:
Another thought, lifted from early conversations on this topic -- Should we add `Msg('section', ...)` which plans can use to declare sections useful for progress bars and other consumers? One path might be to say, "This is a thing we know can do and probably want to do someday," but leave it out of scope for this PR.

Apologies for the stream-of-consciousness posts.

awalter-bnl commented:
> But a visualization, whether a simple static plot or an interactive GUI, can be derived from the low-level generator

I am not sure this is true, as I can see that it would also want to include 'runs', or other possible 'arrangments' and so to properly produce the visualization I mentioned above we would need to run through the messages twice, once in `est_time` and once in `est_time_per_run`. As this information already exists from the `est_time` call inside `est_time_per_run` then I am thinking we should output it as well. I think this approach makes the visualization much easier, as one call gives all of the required information in a list where the positions corresponds to the completion order.

> A intermediate text-based representation looks like it will be quite verbose and hard to parse for realistic plans

I don't think I agree with this, I see the parsing to be quite simple. For example to parse out the cumulative estimate from the yielded value, `val`, from `est_time_per_run` I would use:
`val.cumulative.est_time or val.cumulative.std_dev`
`val.name` would then be an identifying string that can be used in visualizations/ prints and `val.section.est_time/val.section.std_dev` could be used to get the total time for the visualizations/ prints or for further processing.

To be clear, I don't think we should be providing 'every' group and action time in `est_time_per_run`, just those that occur between runs. Everything inside the runs will be contained within the 'run' time, as is proposed to be done for groups in `est_time`.  I think that this resolves the issue of the verbosity of this 'intermediate' approach. We could also limit it to 'actions' that we know take significant time and leave out those that are essentially instantaneous.

> I do think it's useful to consider other reduced representations. I mentioned `est_time_per_point ` above -- having in mind the time per Event.

As an example I would think that this kind of plan would provide the above mentioned tuples for each Event and for each 'action' that occurs between events, following the same pattern as for my description of `est_time_per_run`.

danielballan commented:
> I don't think I agree with this, I see the parsing to be quite simple. For example to parse out the cumulative estimate from the yielded value, val, from est_time_per_run I would use:
val.cumulative.est_time or val.cumulative.std_dev [...]

Oh, I didn't mean hard to parse programmatically, I meant hard to visually read on the screen.

> To be clear, I don't think we should be providing 'every' group and action time in est_time_per_run, just those that occur between runs. Everything inside the runs will be contained within the 'run' time, as is proposed to be done for groups in est_time. I think that this resolves the issue of the verbosity of this 'intermediate' approach. 

I strongly dislike the shift in granularity depending on whether we are in a run.

...But I actually think we don't need to settle those discussions in order to proceed.

In order to un-block progress here, how about this: can you get to an implementation of something like [`est_time` in this comment](https://github.com/NSLS-II/bluesky/pull/1048#issuecomment-400036029) but with two revisions derived from our subsequent conversation:

1. The output `EstTime` should always be in cumulative terms.
2. Yield out `msg, est_time_tuple` instead of just `est_time_tuple`.

That would give us all the information, unfiltered, and enable us to write many variations with just one pass through he plan. Observe how `est_time_per_run` becomes simpler to do:

```py
def est_time_per_run(plan):
    "Takes a plan and yields out total (time, std_err) per run."
    for msg, est in est_time(plan):
        if msg.command == 'close_run':
            yield est
```

Your preferred variation is also easy to implement:

```py
def est_time_per_run(plan):
    "Takes a plan and yields out (label, (time, std_err)) per run."
    runs_open = 0
    group_counter = 0
    run_counter = 0
    for msg, est in est_time(b):
        if msg.command == 'open_run':
            runs_open +=1
        elif msg.command == 'close_run':
            runs_open -= 1
            run_counter += 1
            yield f'run_{run_counter}', est
        elif not runs_open and msg.command == 'wait':
            group_counter += 1
            yield f'group_{group_counter}', est
```


Are you satisfied that that implementation for `est_time` is sufficient to support the high-level APIs we need? If so, should we table the discussion of which high-level APIs are useful until we have the low-level `est_time` fully sorted out?

danielballan commented on bluesky/tests/test_est_time.py:
```diff
@@ -0,0 +1,105 @@
+
+
+#set up the telemetry dictionary with some known values. 
+def setup_telemetry():
```
This is a good place to use a pytest fixture.

awalter-bnl replied:
I will look into pytest fixtures.

danielballan commented on bluesky/tests/test_est_time.py:
```diff
@@ -0,0 +1,105 @@
+import pytest
+
+#setup some testing objects
+RE=RunEngine()
```
Instead of sharing instances across the module, make a fresh one per test by using pytest fixtures. This avoids unexpected and hard-to-debug interactions between separate tests.

awalter-bnl replied:
see above

danielballan commented on bluesky/tests/test_est_time.py:
```diff
@@ -0,0 +1,105 @@
+def test_EpicsMotorEstTime():
+    #test the time estimation for epics motorlike devices.
+    setup_telemetry()
+    assert round(motor1.est_time.set(1,2,1,4).est_time,4) == 4.9975
```
The standard way to do float comparisons in tests as `assert_almost_equal` or `assert_allclose`. You can find examples elsewhere in the bluesky tests.

awalter-bnl replied:
I will look at both of these options that I didn't find in my initial investigation of pytest.

awalter-bnl commented:
I am now seeing your point of view regarding having only msg's and groups included in the base `est_time` function. I think our differences now occur in the output structure.

> In order to un-block progress here, how about this: can you get to an implementation of something like est_time in this comment but with two revisions derived from our subsequent conversation:
>1. The output EstTime should always be in cumulative terms.
>2. Yield out msg, est_time_tuple instead of just est_time_tuple

>That would give us all the information, unfiltered, and enable us to write many variations with just one pass through he plan. 

So this does not solve the main problem that we are disagreeing on (at least from what I understand), which is that we _are_ losing information by not also returning the estimated time (or start time) for each part as well as the cumulative stop time. For example: 

- without the est_time for each part we could not make the 'schedule bar-graph' visual representation I mentioned earlier as we would need both _start_ and _stop_ times for each element, not just the stop readout given by cumulative time. This information is, in particular, lost for 'groups' in your proposed structure, but is also not 100% extractable from the proposed output for 'runs' (particularly if we have multiple overlapping 'runs').
 
-  I see having both as also being useful for de-bugging, in that it will be very easy to determine if an action is being missed (by the large difference between stop time of one action and the start time of the next). This indicates that we have an action that is clearly missed in the output of either the base `est_time` function or a higher level function.

So the question that is holding up the progress at present is what output structure do we need/want? As the proposed base `est_time` function also includes 'groups' then we will need to resolve this before working on that. In order to ensure that no information is lost I see two options (I am happy to consider others):

1. We output both the cumulative stop time and the time to complete each element (msg, group, run).
2. We output a cumulative _start_ as well as stop time for each element (msg, group, run).

In terms of making it easier to read and understand the output perhaps 2. is better but I don't have a huge preference either way.

danielballan commented:
I'm not completely following. Let's focus on (1).

> We output both the cumulative stop time and the time to complete each element (msg, group, run).

Help me understand what this means.

* Time per msg: I think you can get this by comparing (subtracting) two subsequent items.
* Time per group: I guess you could say that "time per group" is the time elapsed from the first 'set'/'trigger'/'kickoff'/'complete' that uses a given group label to the time that that group is waited on. This is easy to compute by, again, subtracting the cumulative time of that first labeled message from the time of the wait message. Any consumer of `est_time` can cache that info and compute on the fly, in one pass.
* Time per run: Similarly, I can watch 'open_run' go back, note the cumulative time, and subtract that from the matching 'close_run'.

What information are we missing?

awalter-bnl commented:
> Time per msg: I think you can get this by comparing (subtracting) two subsequent items.

Agreed we do not have issues with time per msg.

> Time per group: I guess you could say that "time per group" is the time elapsed from the first 'set'/'trigger'/'kickoff'/'complete' that uses a given group label to the time that that group is waited on. This is easy to compute by, again, subtracting the cumulative time of that first labeled message from the time of the wait message. Any consumer of est_time can cache that info and compute on the fly, in one pass.

Your version of `est_time` only outputs the cumulative stop time on the 'wait' messages not on every 'set'/'trigger'/'kickoff'/'complete' message. It also does not output this on the first such message, and so this information is lost based on the current structure. We could have it 'yield' on the first message, but then we are confusing the output by saying that 'if it is  a 'set'....... type message it is a 'start time' otherwise it is a 'stop time'. I prefer to include a cumulative start and stop time on each yield to avoid this confusion.

> Time per run: Similarly, I can watch 'open_run' go back by, note the cumulative time, and subtract that from the matching 'close_run'.

In principle this can be done for runs, however I could imagine a version of `est_time_run` where we only output info on each run element. While the msgs and groups that occur between runs would be included in the cumulative stop time of `est_time`, your output of only cumulative time for such a higher level function would loose the information regarding times between runs.

Is this clearer?

danielballan commented:
> Your version of est_time only outputs the cumulative stop time on the 'wait' messages not on every 'set'/'trigger'/'kickoff'/'complete' message.

I don't understand this. The proposed `est_time` would output this:

```py
...  # assume we are showing the middle of some long est_time output
(Msg('close_run'), EstTime(100, 0.001))
(Msg('set', motor1, 3, group='A'), EstTime(100.1, 100.1))  # first usage of group 'A'
(Msg('set', motor2, 3, group='A'), EstTime(100.2, 100.141))
(Msg('wait', None, group='A'), EstTime(113, 3))  # for some slow motor
...
```

So the "group time" for 'A' would be `113 - 100`.

awalter-bnl commented:
So for those playing at home, there was a long discussion between @danielballan and myself this morning trying to resolve the differences of opinions we have on this. We have come up with the following path forward:

- `est_time` will follow the structure that @danielballan suggested above.

- The output will be a named tuple following the structure `(msg, cumulative_time, time_for_msg)` with `cumulative_time` and `time_for_msg` also being named tuples with the structure `(est_time, std_dev)`. Due to the structure of `est_time` a `wait` message returns the time for the group it waits on in `time_for_msg`.

- All higher level `est_time_....` functions will have the same output structure.

I will make these changes in due course.

danielballan commented:
Per our chat, @awalter-bnl means `time_for_msg` to be `time_for_this_unit` (run, wait group, etc.). I have said I still wonder if that aggregation would be better done one layer up, as different consumers of this information might be interested in different aggregates (by sample instead of by run, for example).

I am also not sure that "higher level est_time_.... functions will have the same output structure," as this seems like a good application for low-level and high-level APIs that may have different return types.

But we have better alignment than we did before, and I reiterate these concerns here only to flag points of future discussion. I think @awalter-bnl should implement exactly what he specified above without regard to my tentative concerns, and then we can resume the discussion from there.

awalter-bnl commented:
The latest commit represents a major change in the structure/ API, this has not been tested or been checked for PEP8, this will be done once we have agreement that this is the structure/ API we want to use.

Notes on the new structure/API

-	Basic output structure is now a named tuple of the form `(message, stop_time, start_time)` where.
o	`message` is the full message yielded by `plan`
o	 `stop_time` and `start_time` are named tuples of the form `(est_time, std_dev)`
-	A base level function, with maximum granularity, is now provided via `est_time(plan)`
•	This yields for each `message`, with the yield for `wait` messages being the times for the associated group as a whole
-	Higher level functions reduce the verbosity of `est_time(plan)`
o	Examples: 
•	`est_time_run(plan)` which modifies the `start_time` in the `close_run` yield of `est_time(plan)` to be the start of the associated run. 
•	`est_time_per_run(plan)` which takes `est_time_run(plan)` and removes the yields between `open_run` and `close_run` messages.
•	`est_time_per_group(plan)` which takes `est_time(plan)` and removes the yields for ‘groupable’ messages.
•	Many others that have not been thought of yet !-)
-	Output functions take these ‘times’ and do something with them.
o	Examples:
•	`print_est_time(plan, est_time_func)` which takes the yields from `est_time_func(plan)` and prints them in a human readable format.
•	I would like to make `plot_est_time` which is like `print_est_time` but makes a ‘schedule bar graph’ plot out of them.
o	Getting the exact ouput that Dan wants could also be done using an output function, this is not true in the other direction.
•	Example. `>set(motor1, x , ‘groupA’), set(motor2, y, ‘groupB’), wait( ‘groupB’), set(motor3, z, ‘groupA’)` is a valid construct but in it would not be possible to ‘locate’ the `motor3` move inside `groupA` without having the start and stop time for the third `set` yielded, particularly if it finishes before the `motor1` move.

A few points to discuss:
-	Looking at the structure here I am wondering if it is worth removing the start and stop time for the `group` (yielded on the `wait` msg) from `est_time`. 
o	It could then be done in  a higher level function `est_time_group`, in the same way it is done for runs in `est_time_run`, this would be more consistent. 
-	I like this structure (with both low level and high level functions having the same API) because it makes the ‘Output functions’ generic, giving the user the ability to ‘select’ the verbosity level for a given function.
-	I am still not convinced that ‘communication time’ and ‘internal_state_time’ can be accurately provided. 
o	I don’t think they are ‘static’ and they will change from beamline to beamline and day to day (perhaps even hour to hour).
o	They have been included in case someone comes up with a good solution for estimating them.
-	If we want to include ‘communication_time’ independently into the times defined for messages that are estimated from telemetry (which they are in this draft) then we will need to make significant changes to the status_objects.
o	This is because at present this time is not accounted for in the estimation of the time for the move that is stored in the telemetry. As a result including it again here will mean it is counted twice.

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +120,424 @@ def check_limits(plan):
+        if msg.command in _group_start_cmds:
+            # add set and trigger commands to plan_history
+            if msg.command == 'set':
+                plan_history['set'][msg.obj.name] = msg.args[0]
```
Why pick off just `msg.args[0]` here? What about any others `args` and `msg.kwargs`?

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +120,424 @@ def check_limits(plan):
+        elif msg.command is 'sleep':
+            start_time = out_time
+            out_time = combine_est(out_time, _TimeStats(msg.args[0], 0))
+            out_time = combine_est(out_time, _communication_time)
```
Is this a copy/paste mistake? There are no network calls involved in sleep.

danielballan commented on bluesky/simulators.py:
```diff
@@ -114,3 +120,424 @@ def check_limits(plan):
+# The commands that aren't  groupable but still have statistics.
+_timed_cmds = ['stage', 'unstage']
+# The commands that only require network communication.
+_communication_cmds = ['read', 'describe', 'open_run', 'close_run']
```
'read' and 'describe' are very different from 'open_run' and 'close_run'. The last two are internal state-related.

awalter-bnl commented:
@danielballan all 3 of your comments are valid, however before we get into these 'details' I want to agree that the overall structure and API are what we want. This will avoid me spending a lot of time resolving these small issues only to have to rewrite large portions of the code later anyway when it is decided that the structure and API are not what we want (as happened last time). 

Can I please get feedback on the structure and API first, once this is done we can go over the the details.

awalter-bnl commented:
I have cleaned up the code with flake8

danielballan commented:
Yes. @tacaswell and I had a thought about how to resolve this at SciPy but need some time to write it up. Maybe @tacaswell can get to it before I get back from vacation?

danielballan commented:
Notes from call with @awalter-bnl:

We will go with this API: `est_time(plan) -> yields (processing_time, action_time)` where `processing_time` is always an `EstTime` representing that time it takes to process a given msg and `action_time` is `None` for synchronous messages (open_run, checkpoint, etc.) or another `EstTime` for any messages that return status objects, representing their time to complete

I am bothered by the return type instability of this design --- some message types return `(EstTime, None)` while others return `(EstTime, EstTime)` --- but this is a reflection of the return type instability of what RE `send`s back to the plan for a given Msg (status object or not status object).

Higher level consumers have all the information they need to produce aggregates like "time per run", "time per group," or "which was the slow motor in this group.

It will eventually be important for the various constant times (internal state update time, communication time, document emitting time) to be configurable somehow in case the caller knows things about local configuration, such as how time-costly the callbacks are, that it wants to incorporate. For now we can use guesses --- `0` is a good zeroth-order approximation. Down the road, maybe we can get some data on communication time by doing spot checks with `EchoRequest`.

awalter-bnl commented:
Latest commits, updated the API as described above. Also added tests for each of the est_time functions. In progress new commits to the ophyd (https://github.com/NSLS-II/ophyd/pull/567/files) in order to remove the 'communication time' from the current status object telemetry estimates.
A few important points:
- I have 'renamed' est_time to est_delta_time (if this is not liked then I can change it again suggestions).
- I have written a new ''est_time' which uses 'est_delta_time' and converts to a cumulative start/stop time (as was the previous outputting.
- This allows the rest of the est_time functions and the print_est_time function to remain unchanged.

awalter-bnl commented:
FYI looking into why the tests are failing on Travis when they work fine locally !!

awalter-bnl commented:
OK so the error for all Travis builds relates to the lack of the related ophyd PR (https://github.com/NSLS-II/ophyd/pull/567/files). This should work once that PR is merged.

danielballan commented:
> This should work once that PR is merged.

To test this against the corresponding ophyd changes, you can add a line to `.travis.yml` to install ophyd specifically from that git branch. We will then revert the change before finally merging. We don't do this often, but it is appropriate for large coordinated changes like this one. For example, the last time I recall doing it was for the "asset refactor" work.

awalter-bnl commented:
@danielballan I am reluctant to do this because:
1. The ophyd branch work will need to be merged first anyway (this does the telemetry side of things so I think we should aim for that to be merged, deployed and generating data prior to merging this side of things so we can get some more test data).
2. Given the timeline for how often this is revisited I will definitely forget to undo this change prior to merging and who knows when it will then be noticed.

danielballan commented:
Fine by me.

awalter-bnl commented:
Based on a discussion with @danielballan, and the issues found trying to rebase the ophyd counterpart to the current master, I am closing this off. Potentially a new PR may be opened in the future.
